shared/src/balancer_event_handler.rs

```diff
@@ -7,31 +7,107 @@ use crate::{
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    ops::RangeInclusive,
+};
 use tokio::sync::Mutex;
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 pub enum BalancerEvent {
     PoolRegistered(PoolRegistered),
+    TokensRegistered(TokensRegistered),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct PoolRegistered {
     pub pool_id: H256,
     pub pool_address: H160,
     pub specialization: PoolSpecialization,
 }
 
+#[derive(Debug, Clone)]
+pub struct TokensRegistered {
+    pub pool_id: H256,
+    pub tokens: Vec<H160>,
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct WeightedPool {
+    pool_id: H256,
+    pool_address: Option<H160>,
+    tokens: Option<Vec<H160>>,
+    specialization: Option<PoolSpecialization>,
+    block_created: u64,
+}
+
+impl WeightedPool {
+    fn update_from_event(&mut self, event: BalancerEvent) {
+        // Pool and Token Registration always occur in the same tx! So block_created doesn't need update.
+        // https://github.com/balancer-labs/balancer-v2-monorepo/blob/70843e6a61ad11208c1cfabf5cfe15be216ca8d3/pkg/pool-utils/contracts/BasePool.sol#L128-L130
+        match event {
+            BalancerEvent::PoolRegistered(pool_registration) => {
+                self.pool_address = Some(pool_registration.pool_address);
+                self.specialization = Some(pool_registration.specialization);
+            }
+            BalancerEvent::TokensRegistered(token_registration) => {
+                self.tokens = Some(token_registration.tokens)
+            }
+        }
+    }
+}
+
+impl From<(EventIndex, PoolRegistered)> for WeightedPool {
+    fn from(event_data: (EventIndex, PoolRegistered)) -> Self {
+        Self {
+            pool_id: event_data.1.pool_id,
+            pool_address: Some(event_data.1.pool_address),
+            specialization: Some(event_data.1.specialization),
+            // Tokens not emitted with pool registration event.
+            tokens: None,
+            block_created: event_data.0.block_number,
+        }
+    }
+}
+
+impl From<(EventIndex, TokensRegistered)> for WeightedPool {
+    fn from(event_data: (EventIndex, TokensRegistered)) -> Self {
+        Self {
+            pool_id: event_data.1.pool_id,
+            // pool_address and specialization not emitted with token registration event.
+            pool_address: None,
+            specialization: None,
+            tokens: Some(event_data.1.tokens),
+            block_created: event_data.0.block_number,
+        }
+    }
+}
+
+/// The BalancerPool struct represents in-memory storage of all deployed Balancer Pools
+#[derive(Debug, Default)]
+pub struct BalancerPools {
+    /// Used for O(1) access to all pool_ids for a given token
+    pools_by_token: HashMap<H160, HashSet<H256>>,
+    /// WeightedPool data for a given PoolId
+    pools: HashMap<H256, WeightedPool>,
+    /// Used mainly as a fallback for `last_event_block` when there are no events.
+    contract_deployment_block: u64,
```

Comment:
No big deal, but is this really needed given that we pass deployment information with the contract instance? Couldn't we just default to block 0?

Reply:
We could default to zero. I like this suggestion and will go with it.